app/main.py

```diff
@@ -1,214 +1,36 @@
-
-    # apply potion if exist
-    if mordred["potion"] is not None:
-        if "power" in mordred["potion"]["effect"]:
-            mordred["power"] += mordred["potion"]["effect"]["power"]
-
-        if "protection" in mordred["potion"]["effect"]:
-            mordred["protection"] += mordred["potion"]["effect"]["protection"]
-
-        if "hp" in mordred["potion"]["effect"]:
-            mordred["hp"] += mordred["potion"]["effect"]["hp"]
-
-    # red_knight
-    red_knight = knightsConfig["red_knight"]
-
-    # apply armour
-    red_knight["protection"] = 0
-    for a in red_knight["armour"]:
-        red_knight["protection"] += a["protection"]
-
-    # apply weapon
-    red_knight["power"] += red_knight["weapon"]["power"]
-
-    # apply potion if exist
-    if red_knight["potion"] is not None:
-        if "power" in red_knight["potion"]["effect"]:
-            red_knight["power"] += red_knight["potion"]["effect"]["power"]
-
-        if "protection" in red_knight["potion"]["effect"]:
-            red_knight["protection"] += red_knight["potion"]["effect"]["protection"]
-
-        if "hp" in red_knight["potion"]["effect"]:
-            red_knight["hp"] += red_knight["potion"]["effect"]["hp"]
-
-    # -------------------------------------------------------------------------------
-    # BATTLE:
-
-    # 1 Lancelot vs Mordred:
-    lancelot["hp"] -= mordred["power"] - lancelot["protection"]
-    mordred["hp"] -= lancelot["power"] - mordred["protection"]
-
-    # check if someone fell in battle
-    if lancelot["hp"] <= 0:
-        lancelot["hp"] = 0
-
-    if mordred["hp"] <= 0:
-        mordred["hp"] = 0
-
-    # 2 Arthur vs Red Knight:
-    arthur["hp"] -= red_knight["power"] - arthur["protection"]
-    red_knight["hp"] -= arthur["power"] - red_knight["protection"]
-
-    # check if someone fell in battle
-    if arthur["hp"] <= 0:
-        arthur["hp"] = 0
-
-    if red_knight["hp"] <= 0:
-        red_knight["hp"] = 0
+from app.modules.knight import Knight
+from app.modules.knights import knights as knights_origin_dict
+
+
+def battle(knights: dict) -> dict:
+    # storage of all knights objects in a dictonary
+    knights_dict = {}
+    # parse inner knights dict. In an each iteration adds all knight features
+    for knight_unit in knights.values():
+        knights_dict[knight_unit["name"]] = Knight(
+            knight_unit["name"],
+            knight_unit["power"],
+            knight_unit["hp"]
+        )
+        knights_dict[knight_unit["name"]].apply_armour(knight_unit["armour"])
+        knights_dict[knight_unit["name"]].apply_weapon(knight_unit["weapon"])
+        knights_dict[knight_unit["name"]].apply_potion(knight_unit["potion"])
+
+    # create instance variables for better reading of code
+    lancelot = Knight.knights["Lancelot"]
+    arthur = Knight.knights["Arthur"]
+    mordred = Knight.knights["Mordred"]
+    red_knight = Knight.knights["Red Knight"]
+
+    print(Knight.knights["Lancelot"])
```

Comment:
why do you need this? dlete useless print()

Reply:
Because original code uses printing of lancelot info